nntrainer/utils/profiler.h

```diff
@@ -12,50 +12,91 @@
  */
 #ifndef __PROFILER_H__
 #define __PROFILER_H__
-#ifndef PROFILE
-
-#define START_PROFILE(event_key)
-#define END_PROFILE(event_key)
-#define REGISTER_EVENT(event_key, ret)
-
-#else
-
-#define START_PROFILE(event_key)                         \
-  do {                                                   \
-    auto &prof = nntrainer::profile::Profiler::Global(); \
-    prof.start(event_key);                               \
-  } while (0);
-
-#define END_PROFILE(event_key)                           \
-  do {                                                   \
-    auto &prof = nntrainer::profile::Profiler::Global(); \
-    prof.end(event_key);                                 \
-  } while (0);
-
-#define REGISTER_EVENT(event_str, ret)                   \
-  do {                                                   \
-    auto &prof = nntrainer::profile::Profiler::Global(); \
-    ret = prof.registerEvent(event_str);                 \
-  } while (0);
-
-#endif /** PROFILE */
 
 #include <chrono>
 #include <future>
 #include <iosfwd>
+#include <list>
 #include <mutex>
+#include <set>
 #include <string>
 #include <unordered_map>
 #include <unordered_set>
 #include <vector>
+
+using timepoint = std::chrono::time_point<std::chrono::steady_clock>;
+
+#ifndef PROFILE
+
+#define PROFILE_TIME_START(event_key)
+#define PROFILE_TIME_END(event_key)
+#define PROFILE_TIME_REGISTER_EVENT(event_key, event_str)
+#define PROFILE_MEM_ALLOC(ptr, size, str)
+#define PROFILE_MEM_DEALLOC(ptr)
+
+#else /** PROFILE */
+
+#define PROFILE_TIME_START(event_key) \
+  nntrainer::profile::Profiler::Global().start(event_key)
+
+#define PROFILE_TIME_END(event_key) \
+  nntrainer::profile::Profiler::Global().end(event_key)
+
+#define PROFILE_TIME_REGISTER_EVENT(event_key, event_str)                 \
+  do {                                                                    \
+    event_key =                                                           \
+      nntrainer::profile::Profiler::Global().registerTimeItem(event_str); \
+  } while (0)
+
+#define PROFILE_MEM_ALLOC(ptr, size, str) \
+  nntrainer::profile::Profiler::Global().alloc(ptr, size, str)
+
+#define PROFILE_MEM_DEALLOC(ptr) \
+  nntrainer::profile::Profiler::Global().dealloc(ptr)
+
+#endif /** PROFILE */
+
 namespace nntrainer {
+
 namespace profile {
 
-typedef enum {
-  NN_FORWARD = 1 /**< Neuralnet single inference without loss calculation */,
-} EVENT;
+enum PROFILE_EVENT {
+  EVENT_TIME_START = 0,
+  EVENT_TIME_END = 1,
+  EVENT_MEM_ALLOC = 2,
+  EVENT_MEM_DEALLOC = 3,
+};
```

Comment:
`EVENT_MEM_ALLOC` and `EVENT_MEM_DEALLOC` are newly added but there are no test cases for those events in unittest_nntrainer_profiler.cpp file. What about adding more test cases for those events?

Reply:
Thanks for checking. As your comment, test for memeory alloc/dealloc profiling is missed.
I implemented, please check them.

Reply:
The new patch contains the test cases of memory profiling as below. :100: 
- TEST_F(ProfileTest, memoryTestAlloc_01_p)
- TEST_F(ProfileTest, memoryTestDealloc_01_p)